### ISSUE TITLE
PSPNet - Pyramid Scene Parsing Network

### ISSUE THREAD

**Issue by junxnone** [junxnone/tio#601]
- FCN 不能有效的处理场景之间的关系和全局信息
- 提出 `Pyramid Pooling Module` - 金字塔池化模块, 聚合不同区域的上下文信息
- 在基于深度监督损失ResNet上制定有效的优化策略

# Reference
- [ ] [paper - Pyramid Scene Parsing Network](https://arxiv.org/pdf/1612.01105.pdf)
- [ ] [[pspnet - project](https://hszhao.github.io/projects/pspnet/)
- [ ] [Github repo](https://github.com/hszhao/PSPNet)
- [ ] [Keras implementation - pspnet](https://github.com/Vladkryvoruchko/PSPNet-Keras-tensorflow)
- [ ] [Semantic Segmentation--Pyramid Scene Parsing Network(PSPNet)论文解读](https://blog.csdn.net/u011974639/article/details/78985130)
- [ ] [netscope - PSPNet-101](http://ethereon.github.io/netscope/#/gist/8a5cc8d2e4fe4020caa8d65716fa7cac)

# Brief
- baseline network :  FCN and dilated network
- ResNet + Dilated network 获取 feature map
  - ResNet 初始的 7x7 Conv 替换为 3 个 3x3 Conv
- Pyramid Pooling Module 收集不同层次的特征
- Concatenate `Pyramid Pooling Module` 的输出和原 `feature map`
- Conv 产生最终 `Predict map`
- 使用了辅助loss -  Auxiliary Loss
- 使用 `poly` 学习率变化策略

## Pyramid Pooling Module
常用的获取全局信息的方法：
- GAP
  - 会失去空间信息导致模糊
- 金字塔池化收集不同层次的特征

---
![image](https://user-images.githubusercontent.com/2216970/70215055-9d8c8300-1777-11ea-88e1-dec007b410a3.png)
- 4 种不同的 pyramid scales 得到不同大小的 feature map
  - bin size 1x1, 2x2, 3x3, 6x6
- 1x1 的卷积 降维
- upsample 到 原 feature map 大小
  - bilinear interpolation
- Concatenate feature map

## Test Result
- GAP 比 Max Pooling 效果更好
- Pooling 之后再降维效果更好
- 
---
![image](https://user-images.githubusercontent.com/2216970/70215931-851d6800-1779-11ea-9d2d-38308eafa1fb.png)

## Auxiliary Loss
- 辅助 loss 添加在 res4b22 Block , 训练时和主 loss 一起工作，取较小权重
- 测试时，放弃辅助 loss

---
![image](https://user-images.githubusercontent.com/2216970/70217504-450bb480-177c-11ea-8003-f6df89792cc7.png)
![image](https://user-images.githubusercontent.com/2216970/70217840-e2ff7f00-177c-11ea-91ac-410c7b747c20.png)

---
辅助 loss 权重测试结果
![image](https://user-images.githubusercontent.com/2216970/70219667-2a3b3f00-1780-11ea-9716-e9d9e4163c57.png)
